src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,378 @@
+ *         meant for connecting to an AsynMemIO's request input, and should not be connected to anything else
+ */
+class MemAccessIO extends Bundle {
+  val address  = Input(UInt(32.W))
+  val valid    = Input(Bool())
+  val response = Flipped(Valid(UInt(32.W)))
+
+  val ready    = Output(Bool())
+  val request  = Decoupled(new Request)
+}
+
+/** 
+ * The *interface* of the IMemAccess module.
+ *
+ * Input:  address, the address of an instruction in memory 
+ * Input:  valid, true when the address specified is valid
+ *
+ * Output: instruction, the requested instruction
+ * Output: ready, true when memory is idling and ready for a request
+ */
+class IMemAccessIO extends MemAccessIO {
+  val instruction = Output(UInt(32.W))
+}
+
+/**
+ * The *interface* of the DMemAccess module.
+ *
+ * Input:  address, the address of a piece of data in memory. 
+ * Input:  writedata, valid interface for the data to write to the address
+ * Input:  valid, true when the address (and writedata during a write) specified is valid
+ * Input:  memread, true if we are reading from memory
+ * Input:  memwrite, true if we are writing to memory
+ * Input:  maskmode, mode to mask the result. 0 means byte, 1 means halfword, 2 means word
+ * Input:  sext, true if we should sign extend the result
+ *
+ * Output: readdata, the data read and sign extended
+ * Output: ready, true when memory is idling and ready for a request
+ */
+class DMemAccessIO extends MemAccessIO {
+  val writedata = Input(UInt(32.W))
+  val memread   = Input(Bool())
+  val memwrite  = Input(Bool())
+  val maskmode  = Input(UInt(2.W))
+  val sext      = Input(Bool())
+
+  val readdata  = Output(UInt(32.W))
+}
+
+/**
+ * The instruction memory accessor.
+ *
+ * The I/O for this module is defined in [[IMemAccessIO]].
+ */
+class IMemAccess extends Module {
+  val io = IO (new IMemAccessIO)
+  io := DontCare
+
+  // Per the ready/valid interface spec
+  when (io.valid && io.ready) {
+    val request = Wire(new Request())
+    request := DontCare
+    request.address   := io.address
+    request.operation := 1.U
+
+    io.request.bits  := request
+    io.request.valid := true.B
+  }
+
+  // When the memory is outputting a valid instruction
+  when (io.response.valid) {
+    io.instruction := io.response.bits
+  }
+}
+
+/**
+ * The data memory accessor.
+ *
+ * The I/O for this module is defined in [[DMemAccessIO]].
+ */
+class DMemAccess extends Module {
+  val io = IO (new DMemAccessIO)
+  io := DontCare
+
+  val storedWrite = RegInit(0.U.asTypeOf(new PartialWrite))
+
+  // Per the ready/valid interface spec
+  when (io.valid && io.ready && io.memread =/= io.memwrite) {
+    val request = Wire (new Request)
+    request.address   := io.address
+    request.writedata := 0.U
+    request.operation := 1.U
+
+    io.request.bits  := request
+    io.request.valid := true.B
+
+    when (io.memwrite) {
+      val partialWrite = Wire(new PartialWrite) 
```

Comment:
I don't think you need this wire. I believe you can just say `storedWrite.address := ...`